### ISSUE TITLE
[PROPOSAL] Automate merging of branch version bump PRs

### ISSUE THREAD

**Issue by dbwiddis** [opensearch-project/automation-app#37]
### What are you proposing?

We should enable repo/admin-level automation to merging branch version bump PRs because maintainers aren't doing it.

### What users have asked for this feature?

I requested it in the 2.14.0 retrospective.

I have repeatedly nagged other plugin maintainers to merge upstream dependencies so I can close mine.

I have received pushback from other plugin maintainers for the above nagging, blocking my efforts to actually close open PRs on repos that I maintain.

Other mentions of automation or enforcement of version bumping:
 - https://github.com/opensearch-project/opensearch-build/issues/2706
 - https://github.com/opensearch-project/opensearch-build/issues/2949

### What problems are you trying to solve?

This, on a repo I maintain ([source](https://github.com/opensearch-project/flow-framework/pulls/app%2Fopensearch-trigger-bot)).
![Screenshot 2024-08-08 at 10 46 48 PM](https://github.com/user-attachments/assets/3bc66486-a102-48f3-b5d2-4f5c4122cb99)

And this on an upstream dependency, blocking me from merging mine ([source](https://github.com/opensearch-project/ml-commons/pulls/app%2Fopensearch-trigger-bot)).
![Screenshot 2024-08-08 at 10 47 20 PM](https://github.com/user-attachments/assets/2b9c1c38-7d8a-4fdc-8ca3-514958ff75fb)

And this on another upstream dependency, blocking me from merging mine ([source](https://github.com/opensearch-project/k-nn/pulls/app%2Fopensearch-trigger-bot)).
![Screenshot 2024-08-08 at 10 49 17 PM](https://github.com/user-attachments/assets/c9336fc3-6926-4cbf-a7db-1dce1dd46f70)

Organization-wide, `opensearch-project` [has 204 unmerged version bump PRs](https://github.com/search?q=org%3Aopensearch-project+Increment+version+to++&type=pullrequests&state=open).

These are literally mouse clicks away from being closed, but it takes upstream repos to lead the way.

### What is the developer experience going to be?

For maintainers like me, the ability to merge PRs on their repo because upstream repositories have appropriate versioning.  

For maintainers who don't care about these PRs, they won't have to lift a finger. How awesome is that?

More seriously, minor version bump PRs are assumed to have been cut as part of automation (Autosync) and multiple developers spent multiple hours dealing with the aftermath of a branch cut prior to a version bump in the 2.14 release.  That's at least an hour of my time wasted, multiplied by at least 4 other developers.

#### Are there any security considerations? 

Patch version bumps are one of the biggest culprits here, because patch releases rarely happen.

However, when they do, it's usually for a very important issue that can't wait until the next release cycle.  In this case, having plugins wait for multiple upstream dependencies to merge their patch version bumps could slow our ability to react to these.

#### Are there any breaking changes to the API

Nope.

### What is the user experience going to be?

Seeing plugin repositories that are well-maintained without a [huge backlog of ignored PRs that discourage them from contributing](https://danluu.com/discourage-oss/).

#### Are there breaking changes to the User Experience?

Nope.

### Why should it be built? Any reason not to?

It should be built because automation already exists to create the PRs; they can be auto-merged by a bot with appropriate powers to do so, when all dependencies are met.

It should be built to save the time and effort of maintainers to do the same.  In particular, GitHub action retries expire after 30 days, requiring a minute or so of effort to re-try a version bump PR after a month of upstream repos ignoring it... or similar effort to search and identify whether it's able to be merged. It's a distraction and complete waste of developer time.

### What will it take to execute?

Whatever automation creates the PRs can be given the power to merge them if CI checks are gren.

### Any remaining open questions?

Why don't we just require maintainers to do this?   Actually, we do, in Release Checklists, which specify merging these version bumps as part of the checklist.  There are [453 open Release Checklist issues](https://github.com/search?q=org%3Aopensearch-project+%22Release+Version%22++&type=issues&state=open) and while 3.0.0 and 2.17.0 are a couple hundred of those, there are far more than that.

**Comment by dblock** [comment]
[[Catch All Triage](https://github.com/opensearch-project/.github/blob/main/TRIAGING.md) - [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/andrross/), [4](https://github.com/krisfreedain), [5](https://github.com/amistrn)]

**Comment by gaiksaya** [comment]
I was experimenting with GitHub's auto-merge feature https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request
It looks like a good solution. 

Please note, enabling auto-merge does not mean all the pull requests will merged as soon as the CI is green. It's per pull request based.

```
You can allow setting pull requests to merge automatically once all required reviews and status checks have passed.
```

Below are the 2 approaches:
### 1. Use same workflow to enable auto-merge 
The [same workflow](https://github.com/opensearch-project/opensearch-build/blob/main/.github/workflows/os-increment-plugin-versions.yml) that creates the pull request can enable the auto merge feature too. https://cli.github.com/manual/gh_pr_merge

Command to use:
```
gh pr merge <url>/<number> --auto --squash
```

**Action items to follow:**
1. Enable `auto-merge` in the repo settings that allows the maintainer or users with write access to enable auto-merge for PR.
2. Add the above command to the workflow to enable auto-merge

**Pros:**
- No need to maintain another workflow
- Tightly coupled with existing automation
- Minimal change

**Cons:**
- Blast radius increases. If the workflow has issues (for example, some action is failing) the overall workflow will be failing
- Tightly couples version increment and PR merge 
- Need another solution to manage dependency tree

### 2. Use metrics data to enable auto-merge for all PRs
We also indexing all this data into opensearch metrics cluster https://metrics.opensearch.org
Another approach is to fetch the data and run similar gh commands for those.

**Action items to follow:**
1. Enable `auto-merge` in the repo settings that allows the maintainer or users with write access to enable auto-merge for PR.
2. Create a jenkins library to fetch data and process it
3. Add the above command to the workflow to enable auto-merge via jenkins workflow

**Pros:**
- Separate workflows so no overlap in the blast radius
- Both actions (creating a pull request and merging the existing ones) can be managed independently 
- Can be expanded to check the existence of mandatory plugins such as job-scheduler, common-utils, remote-cdk, etc. 

**Cons:**
- Additional management on jenkins workflow/libraries
- Depends heavily on metrics data. Missing data = no action

Need your opinion and feedback on this please @prudhvigodithi @dbwiddis @peterzhuamazon @rishabh6788 @getsaurabh02 

Thanks!

**Comment by dbwiddis** [comment]
> Need your opinion and feedback on this please

I like both options!   

Option 2 seems a more elegant solution but I'm not sure this really needs such a high tech solution.  Most of the problems could probably be solved with a daily run that merges those with passing CI (or as you indicate, auto-merges) and rebases/retries everything else.   Even multi-layered dependency chains would typically get resolved within a week or so with this strategy. 

The key here is recognizing which repo(s) are the blockers if the process ends up stalled for some reason, and require manual intervention, and triggering appropriate alerts to that repo's maintainers.

Either way I'm happy to see progress toward some solution other than waiting on maintainers!

Dan

**Comment by peterzhuamazon** [comment]
I would agree with Dan here that second option is more accurate and related to each repo's state based on dependency tree. Please the trigger can be optimized based on data.

Thanks.

**Comment by gaiksaya** [comment]
Thank you for your inputs @peterzhuamazon @dbwiddis 
Will add some acceptance criteria to this issue to take it to completion.
